src/net/larse/lcms/algorithms/VCT.java

```diff
@@ -0,0 +1,1567 @@
+     * QA_GOOD values. Implements section 3.3.1 in Huang et al. (2010) paper
+     */
+    private void interpolationAndIndices() {
+
+      // Flag each year's pixel as good or bad based on the mask value
+      // TODO: Note that class 0 (BACKGROUND) is *not* being flagged as 
+      // "fillable" in the original source code (ie. SLC-off errors).  
+      // We need to ask Cheng about this
+      for (int i = 0; i < this.numYears; i++) {
+        if (this.mask[i] != 0 && this.mask[i] <= FILL_CLASSES) {
+          this.qFlag[i] = QA_BAD;
+        } else {
+          this.qFlag[i] = QA_GOOD;
+        }
+      }
+
+      // Look for spikes and dips that may be unflagged cloud and shadow and
+      // set the QA flag accordingly
+      for (int i = 1; i < this.numYears - 1; i++) {
+
+        // Skip over pixels already labeled as QA_BAD
+        if (this.qFlag[i] == QA_BAD) {
+          continue;
+        }
+
+        // Pixels higher in UD than neighboring years - cloud
+        if (isRelativeCloud(i)) {
+          this.qFlag[i] = QA_BAD;
+          continue;
+        }
+
+        // Pixels lower than UD than neighboring years - shadow
+        if (isRelativeShadow(i)) {
+          this.qFlag[i] = QA_BAD;
+        }
+      }
+
+      // Now test the start/end conditions
+      int n = this.numYears;
+      if (this.qFlag[0] == QA_GOOD && isBadEndpoint(0, 1)) {
+        this.qFlag[0] = QA_BAD;
+      }
+      if (this.qFlag[n - 1] == QA_GOOD && isBadEndpoint(n - 1, n - 2)) {
+        this.qFlag[n - 1] = QA_BAD;
+      }
+
+      // Calculate bad percentage
+      int badCount = 0;
+      for (int i = 0; i < this.numYears; i++) {
+        if (this.qFlag[i] == QA_BAD) {
+          badCount += 1;
+        }
+      }
+      this.pctGoodObs = (int) (100.0 - ((100.0 * badCount) / this.numYears));
+
+      // Interpolate for bad observations indicated by qFlag
+      int i = 0;
+      if (this.pctGoodObs > 50.0) {
+        while (i < this.numYears) {
+          // Skip good observations
+          if (this.qFlag[i] == QA_GOOD) {
+            i += 1;
+          }
+          // Fill or interpolate bad observations
+          else {
+            // Search forward/backward to find next valid observations in 
+            // time series
+            int prev = i - 1;
+            int next = i + 1;
+            while (prev >= 0 && this.qFlag[prev] == QA_BAD) {
+              prev -= 1;
+            }
+            while (next < this.numYears && this.qFlag[next] == QA_BAD) {
+              next += 1;
+            }
+
+            // No valid QA_GOOD pixels in the time series
+            if (prev < 0 && next >= this.numYears) {
+              break;
+            }
+            // No acceptable previous QA_GOOD - use next index to fill
+            // all years from 0 to next
+            else if (prev < 0) {
+              for (int j = 0; j < next; j++) {
+                for (int k = 0; k < N_BANDS; k++) {
+                  this.ud[k][j] = this.ud[k][next];
```

Comment:
You can switch the order of these two for() loops so that k is on the outside, and then get rid of the j loop with
    Arrays.fill(this.ud[k], 0, next, this.ud[k][next]);

Same below.